### PR TITLE
Fix softmax issues

### PR DESCRIPTION
**Pull request by minhitbk**
# Description
1. Fix issues wrt the softmax problem when dealing with general behaviour of a classifer's output:
- PytorchClassifier.
- TFClassifier.
2. Remove casting support of float32 type inside PytorchClassifier. This change requires users must ensure input type and weight type in their model must be same.
- This causes all unit tests of all attacks failed + unittest of pixelcnn defense + unittest of metrics also failed. All fixed.
- This causes bugs in about 10 attacks. All fixed.

## Type of change
Please check all relevant options.

- [x] Bug fix (non-breaking)
- [ ] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: tests/attacks/test_adversarial_patch.py [hunk 1]
```diff
@@ -97,15 +97,16 @@ def test_ptclassifier(self):
         """
         ptc = get_classifier_pt()
 
-        x_train = np.swapaxes(self.x_train, 1, 3)
+        x_train = np.swapaxes(self.x_train, 1, 3).astype(np.float32)
```
Comment (beat-buesser) [inline tests/attacks/test_adversarial_patch.py]: Is there a reason for using `np.float32` instead of `NUMPY_DTYPE`?

File: tests/attacks/test_boundary.py [hunk 2]
```diff
@@ -142,7 +142,7 @@ def test_ptclassifier(self):
         # Build PyTorchClassifier
         ptc = get_classifier_pt()
 
-        x_test = np.swapaxes(self.x_test, 1, 3)
+        x_test = np.swapaxes(self.x_test, 1, 3).astype(np.float32)
```
Comment (beat-buesser) [inline tests/attacks/test_boundary.py]: Is there a reason for using np.float32 instead of NUMPY_DTYPE?

File: tests/attacks/test_boundary.py [hunk 3]
```diff
@@ -258,7 +258,7 @@ def test_iris_tf(self):
     def test_iris_pt(self):
         classifier = get_iris_classifier_pt()
         attack = BoundaryAttack(classifier, targeted=False, max_iter=10)
-        x_test_adv = attack.generate(self.x_test)
+        x_test_adv = attack.generate(self.x_test.astype(np.float32))
```
Comment (beat-buesser) [inline tests/attacks/test_boundary.py]: Is there a reason for using np.float32 instead of NUMPY_DTYPE?

File: tests/attacks/test_carlini.py [hunk 4]
```diff
@@ -160,7 +160,7 @@ def test_ptclassifier(self):
         # Build PyTorchClassifier
         ptc = get_classifier_pt()
 
-        x_test = np.swapaxes(self.x_test, 1, 3)
+        x_test = np.swapaxes(self.x_test, 1, 3).astype(np.float32)
```
Comment (beat-buesser) [inline tests/attacks/test_carlini.py]: Is there a reason for using np.float32 instead of NUMPY_DTYPE?

File: tests/attacks/test_carlini.py [hunk 5]
```diff
@@ -447,7 +447,7 @@ def test_ptclassifier(self):
...
```
Comment (beat-buesser) [inline tests/attacks/test_carlini.py]: Is there a reason for using np.float32 instead of NUMPY_DTYPE?

File: tests/attacks/test_carlini.py [hunk 6]
```diff
@@ -560,7 +560,7 @@ def test_iris_tf(self):
     def test_iris_pt(self):
         classifier = get_iris_classifier_pt()
         attack = CarliniLInfMethod(classifier, targeted=False, max_iter=10, eps=0.5)
-        x_test_adv = attack.generate(self.x_test)
+        x_test_adv = attack.generate(self.x_test.astype(np.float32))
```
Comment (beat-buesser) [inline tests/attacks/test_carlini.py]: Is there a reason for using `np.float32` instead of `NUMPY_DTYPE`?

File: tests/attacks/test_elastic_net.py [hunk 7]
```diff
@@ -222,17 +222,19 @@ def test_ptclassifier(self):
         # Build PyTorchClassifier
         ptc = get_classifier_pt()
 
-        x_test = np.swapaxes(self.x_test, 1, 3)
+        x_test = np.swapaxes(self.x_test, 1, 3).astype(np.float32)
```
Comment (beat-buesser) [inline tests/attacks/test_elastic_net.py]: Is there a reason for using `np.float32` instead of `NUMPY_DTYPE`?

File: tests/attacks/test_virtual_adversarial.py [hunk 8]
```diff
@@ -213,20 +221,21 @@ def test_iris_k_unbounded(self):
     #     acc = np.sum(preds_adv == np.argmax(y_test, axis=1)) / y_test.shape[0]
     #     logger.info('Accuracy on Iris with VAT adversarial examples: %.2f%%', (acc * 100))
 
-    def test_iris_pt(self):
```
Comment (beat-buesser) [inline tests/attacks/test_virtual_adversarial.py]: Why did you comment `test_iris_pt`?

Comment (minhitbk) [reply tests/attacks/test_virtual_adversarial.py]: The test_iris_pt and the test_iris_tf get the same error as the attack produces "nan" in both cases as the adversarials. So there maybe problem with the attack itself or the supportability of the attack for vector input. This needs to be investigated more.
